Adds a maximum message size for RDP messages and adds the tdp Notification message

ibeckermayer opened the pull request:
Fixes https://github.com/gravitational/teleport-private/issues/178, including the more general problem discovered and documented in https://github.com/gravitational/teleport-private/issues/178#issuecomment-1299185326

The rust changes all correspond to https://github.com/gravitational/teleport-private/issues/178#issuecomment-1299206750.

The changes in Go add a `Notification` message to TDP, allowing us to send a non-fatal error (warning). The corresponding frontend updates make it so that the non-fatal error shows up as a warning modal. Relates to https://github.com/gravitational/teleport/issues/13041

<img width="2556" alt="Screenshot 2022-11-03 at 12 43 20" src="https://user-images.githubusercontent.com/13578537/199821410-77254caa-478c-4d26-91a4-42e876a0fe4a.png">

As written, a program could still DoS an individual session by causing that modal to continually pop up. This will be solved in an upcoming webapps PR, which will make warnings appear in a dropdown list rather than as a modal.

Corresponds with https://github.com/gravitational/webapps/pull/1333.

#### TDP versioning

On Teleport major version releases, we instruct customers to update proxy/auth before w_d_s, meaning that we can add TDP messages that are sent from server (w_d_s) to client (proxy) without issue (as is the case here). For this reason, these changes won't be backported, and will just be included in the Teleport 12 release.

##### Update

https://github.com/gravitational/teleport/pull/18463 was merged into this PR, see its description for a description of additional changes.

Commit:
Adds a maximum capacity to the vchan::Client to prevent memory DoS attacks

Commit:
Adds tdp.IsNonFatalErr which is used to check for non fatal tdp errors
(as of now, this just means clipDataMaxLenErr). This allows callers of
tdp.Conn functions to optionally respond to TDP errors non-fatally, which
is what we want for the clipDataMaxLenErr error.

Commit:
Updates to include a fatal flag in the tdp Error message

Commit:
Merge branch 'master' into isaiah/fix-TEL-Q322-3

zmb3 commented:
Hmm, could we work with @xinding33 and @rishibarbhaya-design to find a better approach for showing non-fatal errors? Stealing focus with a modal seems kind of annoying - maybe a tooltip in the corner or some sort of warning indicator in the top menu bar would make more sense.

zmb3 commented on lib/srv/desktop/rdp/rdpclient/src/vchan.rs:
```diff
@@ -26,25 +26,34 @@ use std::io::{Cursor, Read};
 pub struct Client {
+    // capacity is effectively the maximum size of an RDP
+    // message we will receive from the RDP server, minus any
+    // ChannelPDUHeader's.
```
```suggestion
    // ChannelPDUHeaders.
```

zmb3 commented on lib/srv/desktop/rdp/rdpclient/src/vchan.rs:
```diff
@@ -26,25 +26,34 @@ use std::io::{Cursor, Read};
+///
+/// To prevent memory DoS attacks that could take down the entire system running the Windows Desktop Service,
+/// Client drops all RDP messages > capacity.
+/// See https://github.com/gravitational/teleport-private/issues/178#issuecomment-1299185326 for context.
```
I wouldn't link to a private issue in OSS code.

zmb3 commented on rfd/0037-desktop-access-protocol.md:
````diff
@@ -200,11 +200,13 @@ This message contains a mouse wheel update. Sent from client to server.
 
 ```
-| message type (9) | message_length uint32 | message []byte
+| message type (9) | message_length uint32 | message []byte | fatal bool
````
Any compatibility issues here? We need to support the old message too since w_d_s may be older than the proxy.

ibeckermayer replied:
(Ah yes, I knew there was something I was forgetting to add to the description). 

I was going to write that it should work fine, because the first section of the message structure remains the same and in practice we only ever decode Errors on the frontend, whose client would just ignore the additional `fatal`. But now that I'm double checking, I realized that we're calling `decodeClipboardData` now from the proxy (via `ReadRaw()`), which will error if the message is in the old format.

Will experiment and come up with a way to account for that.

zmb3 commented on lib/srv/desktop/tdp/conn.go:
```diff
@@ -119,3 +119,14 @@ func (c *Conn) LocalAddr() net.Addr {
+		return false
+	}
+
+	return err.Error() == clipDataMaxLenErr
```
Use `errors.Is` here instead of comparing strings?

zmb3 reviewed (commented):
I still need to catch up on this issue, but a few comments.

Commit:
Update lib/srv/desktop/rdp/rdpclient/src/vchan.rs

Co-authored-by: Zac Bergquist <zac.bergquist@goteleport.com>

Commit:
- reverts maxClipboardDataLength to its proper value
- replaces err.Error() == with errors.Is in IsNonFatalErr
- Adds a TestIsNonFatalErr

Commit:
Swaps out the updated tdp Error changes for a new message Error2

Commit:
swaps out tdp Error2 for Notification with severity

Commit:
Merge branch 'master' into isaiah/fix-TEL-Q322-3

zmb3 commented:
@ibeckermayer let's change the title of this PR (and the resulting commit message) to be more descriptive of what the change is.

"Fixes TEL-Q322-3" is something that should be in the body of the message, but as a title it tells the reader very little.

zmb3 commented on rfd/0037-desktop-access-protocol.md:
````diff
@@ -200,11 +210,28 @@ This message contains a mouse wheel update. Sent from client to server.
+| message type (28) | message_length uint32 | message []byte | severity byte
+```
+
+This message sends a notification message with a severity level.
````
Let's mention that it's sent from Teleport to the browser.

zmb3 commented on rfd/0037-desktop-access-protocol.md:
```diff
@@ -200,11 +210,28 @@ This message contains a mouse wheel update. Sent from client to server.
+This message sends a notification message with a severity level.
+
+`severity` defines the severity of the `message`:
+- `0` is for an error
```
0 for error just seems odd. 0 indicates success in almost every protocol/API I can think of. Can  we skip over 0 and make 1 error, 2 warning, 3 informational ?

ibeckermayer replied:
I don't want to skip 0 because we use it in other enums i.e. in the mouse wheel scroll message
```
`axis` identifies which axis the scroll happened on:
- `0` is vertical scroll
- `1` is horizontal scroll
```

How about 0 info, 1 warning, 2 error?

zmb3 commented on lib/web/desktop.go:
```diff
@@ -353,8 +371,17 @@ func proxyWebsocketConn(ws *websocket.Conn, wds net.Conn) error {
 				return
-			}
-			if err != nil {
+			} else if tdp.IsNonFatalErr(err) {
```
Can we simplify these 2 branches that are almost identical?

```
isFatal := !tdp.IsNonFatalErr(err)
sendTDPNotification(ws, err, isFatal)
```

ibeckermayer replied:
https://github.com/gravitational/teleport/blob/700fff47d6f25fbc974b6ce6618b899b28cbbd4f/lib/web/desktop.go#L372-L385

zmb3 commented on lib/web/desktop.go:
```diff
@@ -333,6 +336,21 @@ func proxyWebsocketConn(ws *websocket.Conn, wds net.Conn) error {
 	}
 
+	sendTDPNotification := func(ws *websocket.Conn, err error, fatal bool) error {
+		var severity tdp.Severity
+		if fatal {
+			severity = tdp.SeverityError
+		} else {
+			severity = tdp.SeverityWarning
+		}
```
```suggestion
		severity := tdp.SeverityError
		if !fatal {
			severity = tdp.SeverityWarning
		}
```

Pick a default so we can kill the else.

zmb3 commented on lib/srv/desktop/tdp/proto_test.go:
```diff
@@ -200,3 +200,15 @@ func TestMFA(t *testing.T) {
 	require.Equal(t, respWant, respGot)
 }
+
+func TestIsNonFatalErr(t *testing.T) {
```
Test a fatal error too? You can generate a slice that's too large with `bytes.Repeat`.

ibeckermayer replied:
1d3851cddba0b9f236d4e410c2c9acd37dc3668f

zmb3 commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -513,14 +550,21 @@ func (c ClipboardData) Encode() ([]byte, error) {
-		return nil, trace.BadParameter("clipboard data exceeds maximum length")
+		// If clipboard data exceeds maxLen,
+		// discard the rest of the message
+		if _, err := io.CopyN(io.Discard, in, int64(length)); err != nil {
```
Do we need this check? Can we just skip straight to returning the max length error? If we don't, and someone sends a clipboard data message with a length field that is too large, but they don't actually send that much data, we'll return an EOF error instead of the max data error.

Additionally, would `trace.LimitExceeded` be better than `BadParameter` here?

zmb3 commented on rfd/0037-desktop-access-protocol.md:
````diff
@@ -200,11 +210,28 @@ This message contains a mouse wheel update. Sent from client to server.
+#### 28 - notification
+
+```
+| message type (28) | message_length uint32 | message []byte | severity byte
````
message_length includes the severity byte, right? or is it the length of the string?

Let's specify since it's unclear what you mean by "message" here - it could be the "notification message" or it could be the "message field" in the "notification message"

ibeckermayer replied:
It's just the length of the string. We're not entirely consistent within this RFD, but in 0067 I began always calling the length of `<name> []byte` fields `<name>_length`. E.g. see

```
| message type (21) | completion_id uint32 | directory_id uint32 | path_length uint32 | path []byte | offset uint64 | write_data_length uint32 | write_data []byte |
```

I'll add a note here to clarify.

zmb3 commented on lib/srv/desktop/rdp/rdpclient/src/vchan.rs:
```diff
@@ -26,25 +26,33 @@ use std::io::{Cursor, Read};
 /// for more information about chunks.
+///
+/// To prevent memory DoS attacks that could take down the entire system running the Windows Desktop Service,
+/// Client drops all RDP messages > capacity.
```
1. Capacity here is an internal implementation detail, so I wouldn't mention it at the Client level. It should be mentioned int he docs for the `new` fn.
2. You don't have to say "to prevent memory DoS attacks that could take down the whole system." Just say "client will drop all messages with length greater than the specified capacity.

zmb3 commented on lib/srv/desktop/rdp/rdpclient/src/lib.rs:
```diff
@@ -1900,3 +1900,5 @@ pub(crate) type Messages = Vec<Message>;
     fn encode(&self) -> RdpResult<Message>;
 }
+
+const MAX_RDP_MESSAGE_SIZE: usize = 2_usize.pow(21); // 2MB
```
The name here suggests that the RDP protocol has a maximum message size, but this is more of an arbitrary value that we selected, so let's use a better name here.

zmb3 commented on lib/srv/desktop/rdp/rdpclient/src/vchan.rs:
```diff
@@ -26,25 +26,33 @@ use std::io::{Cursor, Read};
-    fn default() -> Self {
-        Self::new()
-    }
+    drop_this_message: bool,
```
This seems like an odd name for a field. What message is "this message" supposed to be referring to?

ibeckermayer replied:
I changed it to `drop_current_message`. It's a flag that keeps track of whether or not we've gone over `capacity` for the RDP message currently being read. We need some state to keep track of this, because `Client.read()` might be called several times for a single RDP message ("`current_message`").

Commit:
Stylistic refactoring from code review

Commit:
adds more checks to TestIsNonFatalErr

Commit:
Merge branch 'master' into isaiah/fix-TEL-Q322-3

Commit:
Changes the notification severity enum numbering

Commit:
Adds a `tdpMaxFileReadWriteLength` and tests to ensure size limit violations are non-fatal errors (#18463)

Commit:
Merge remote-tracking branch 'origin/master' into isaiah/fix-TEL-Q322-3

jakule commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -1460,3 +1470,14 @@ func writeUint64(b *bytes.Buffer, v uint64) {
+// (see https://github.com/gravitational/teleport/issues/14950)
+const tdpMaxPathLength = 10240
+
+const maxClipboardDataLength = 1024 * 1024    // 1MB
```
Is this MB or KB?

ibeckermayer replied:
1MB == 2^20 bytes == `0x100000` == (1KB)^2 == 1024 * 1024 bytes

jakule replied:
Yes... 🤦

jakule commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -1460,3 +1470,14 @@ func writeUint64(b *bytes.Buffer, v uint64) {
+// the browser (Teleport never receives this message, so won't be decoding it)
+const tdpMaxNotificationMessageLength = 10240
+
+// tdpMaxPathLength is currently just an arbitrary value, more research is required
```
Should we add TODO here?

jakule commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -1424,14 +1431,17 @@ func encodeString(w io.Writer, s string) error {
 
 	if length > maxLen {
-		return "", trace.BadParameter("TDP string length exceeds allowable limit")
+		io.CopyN(io.Discard, r, int64(length))
```
So it's clear that it's intentional
```suggestion
		_, _ = io.CopyN(io.Discard, r, int64(length))
```

jakule commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -1304,6 +1306,11 @@ func decodeSharedDirectoryWriteRequest(in byteReader) (SharedDirectoryWriteReque
 	}
 
+	if writeDataLength > maxLen {
+		io.CopyN(io.Discard, in, int64(writeDataLength))
```
So, it's clear that is intentional.
```suggestion
		_, _ = io.CopyN(io.Discard, in, int64(writeDataLength))
```

zmb3 commented on lib/srv/desktop/tdp/conn.go:
```diff
@@ -117,3 +118,17 @@ func (c *Conn) LocalAddr() net.Addr {
+		return false
+	}
+
+	return errors.Is(err, trace.LimitExceeded(clipDataMaxLenErr)) ||
```
Are there limit exceeded errors that are fatal?

If not, how about just `return trace.IsLimitExceeded(err)`.

ibeckermayer replied:
I think it's better to limit this to specific error messages so that we don't have to always remember that `trace.LimitExceeded` will be counted as non-fatal.

zmb3 commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -463,13 +462,48 @@ func (m Error) Encode() ([]byte, error) {
+	SeverityError   Severity = 2
+)
+
+type Notification struct {
```
```suggestion
// Notification is an informational message sent from Teleport
// to the Web UI. It can be used for fatal errors or non-fatal
// warnings.
type Notification struct {
```

Add godocs

zmb3 commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -463,13 +462,48 @@ func (m Error) Encode() ([]byte, error) {
+func (m Notification) Encode() ([]byte, error) {
+	buf := new(bytes.Buffer)
+	buf.WriteByte(byte(TypeNotification))
+	if err := encodeString(buf, m.Message); err != nil {
```
Looks like the Web UI counterpart has already merged, so feel free to ignore this suggestion, but I wonder if we should stop creating new messages without a length field included ahead of time.

The pattern we've been following so far makes it nearly impossible to skip messages you don't know about and makes backwards compatibility a pain.

ibeckermayer replied:
I see how that could help us avoid some compatibility problems -- we can just drop messages whose opcodes we're unaware of.

I think its still a bit of a dangerous game though -- for example, if we do an update like this one, where we're now sending `tdp.Notification` rather than `tdp.Error`, then we'd run into the problem of older clients behaving weirdly because they're ignoring error messages.

zmb3 commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -513,14 +545,19 @@ func (c ClipboardData) Encode() ([]byte, error) {
 	return buf.Bytes(), nil
 }
 
+const clipDataMaxLenErr = "clipboard sync failed: clipboard data exceeded maximum length"
```
If you take my `trace.IsLimitExceeded` advice above, you can inline these messages and don't have the need for separate constants.

Also, two minor nits:

- the `Err` suffix on the variable name suggest this is an `error` but it's actually a string
- You have some inconsistent naming - here the suffix is `Err` but below you have mfaDataMaxLenError` with `Error` as a suffix

zmb3 commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -1418,6 +1471,17 @@ func writeUint64(b *bytes.Buffer, v uint64) {
+// the browser (Teleport never receives this message, so won't be decoding it)
+const tdpMaxNotificationMessageLength = 10240
+
+// tdpMaxPathLength is currently just an arbitrary value, more research is required
```
An arbitrary value for _what_ though?

(I assume this is the lenght of a file path, but I would say that in the comment)

jakule commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -1236,6 +1233,11 @@ func decodeSharedDirectoryReadResponse(in io.Reader) (SharedDirectoryReadRespons
 	}
 
+	if readDataLength > maxLen {
+		io.CopyN(io.Discard, in, int64(readDataLength))
```
Explicitly intentional: 
```suggestion
		_, _ = io.CopyN(io.Discard, in, int64(readDataLength))
```

zmb3 commented on lib/srv/desktop/tdp/proto_test.go:
```diff
@@ -200,3 +188,137 @@ func TestMFA(t *testing.T) {
+		msg  Message
+	}{
+		{
+			name: "rejects long ClientUsername",
```
Shouldn't this case be fatal though? The length of a username is capped by Windows, so if this happens, it's not legitimate and there's no need for us to try and continue the session.

ibeckermayer replied:
Good point

zmb3 commented on lib/web/desktop.go:
```diff
@@ -104,7 +104,10 @@ func (h *Handler) createDesktopConnection(
 
 	sendTDPError := func(ws *websocket.Conn, err error) error {
```
Do we need to maintain both `sendTDPError` and `sendTDPNotification`, or can a fatal notification take the place of `sendTDPError`?

zmb3 commented on lib/web/desktop.go:
```diff
@@ -354,8 +370,17 @@ func proxyWebsocketConn(ws *websocket.Conn, wds net.Conn) error {
-			}
-			if err != nil {
+			} else if err != nil {
+				isFatal := !tdp.IsNonFatalErr(err)
```
I wonder if we should refactor and make this `tdp.IsFatalErr`. For me it's much easier to think about than having to deal with the inverted form.

ibeckermayer replied:
I went with
```go
// IsNonFatalErr returns whether or not an error arising from
// the tdp package should be interpreted as fatal or non-fatal
// for an ongoing TDP connection.
func IsNonFatalErr(err error) bool {
	if err == nil {
		return false
	}

	return errors.Is(err, trace.LimitExceeded(clipDataMaxLenErrMsg)) ||
		errors.Is(err, trace.LimitExceeded(stringMaxLenErrMsg)) ||
		errors.Is(err, trace.LimitExceeded(fileReadWriteMaxLenErrMsg)) ||
		errors.Is(err, trace.LimitExceeded(mfaDataMaxLenErrMsg))
}

// IsFatalErr returns the inverse of IsNonFatalErr
// (except for if err == nil, for which both functions return false)
func IsFatalErr(err error) bool {
	if err == nil {
		return false
	}

	return !IsNonFatalErr(err)
}
```
and we can use whichever causes less double negatives in a particular context.

jakule commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -640,7 +636,8 @@ func DecodeMFA(in byteReader) (*MFA, error) {
 
 	if length > maxMFADataLength {
-		return nil, trace.BadParameter("mfa challenge data exceeds maximum length")
+		io.CopyN(io.Discard, in, int64(length))
```
Ignored intentionally:
```suggestion
		_, _ = io.CopyN(io.Discard, in, int64(length))
```

zmb3 commented on rfd/0037-desktop-access-protocol.md:
````diff
@@ -200,15 +210,37 @@ This message contains a mouse wheel update. Sent from client to server.
+| message type (28) | message_length uint32 | message []byte | severity byte |
+```
+
+This message sends a notification message with a severity level. Sent from server to client.
````
```suggestion
This message sends a notification message with a severity level. Sent from Teleport to browser.
```

I don't like using server client here, because it's confusing. The "server" in this case is Teleport, but when it comes to RDP Teleport is the RDP client..

ibeckermayer replied:
I don't agree that RDP is relevant here because this is a TDP specification document. It means "sent from TDP server to TDP client".

ibeckermayer commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -1418,6 +1471,17 @@ func writeUint64(b *bytes.Buffer, v uint64) {
+const tdpMaxPathLength = 10240
+
+const maxClipboardDataLength = 1024 * 1024    // 1MB
+const tdpMaxFileReadWriteLength = 1024 * 1024 // 1MB
```
These aren't exactly the same limits. The limit in the Rust code is for a message received on a vchannel -- it's 2MB because that let's us receive 1MB of raw read/write data with plenty of wiggle room for the additional message fields and headers. This limit is for just the raw read/write data.

Commit:
CR

Commit:
Merge branch 'master' into isaiah/fix-TEL-Q322-3

zmb3 commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -440,19 +444,21 @@ func (r ClientUsername) Encode() ([]byte, error) {
 	return ClientUsername{Username: username}, nil
 }
 
+// Error is a tdp Error. It has been deprecated
```
Is this going to be backported to v11 or only v12 and up?

If we know it's only v12 and up, maybe we mention that here?

```
// Error is used to send a fatal error message to the browser.
// In Teleport 12 and up, Error is deprecated and Notification
// should be preferred.
```

ibeckermayer replied:
This is only 12 and up

> On Teleport major version releases, we instruct customers to update proxy/auth before w_d_s, meaning that we can add TDP messages that are sent from server (w_d_s) to client (proxy) without issue (as is the case here). For this reason, these changes won't be backported, and will just be included in the Teleport 12 release.

zmb3 commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -578,6 +625,8 @@ func (m MFA) Encode() ([]byte, error) {
 	return buf.Bytes(), nil
 }
 
+const mfaDataMaxLenErrMsg = "mfa challenge data exceeds maximum length"
```
```suggestion
const mfaDataMaxLenErrMsg = "MFA challenge data exceeds maximum length"
```

zmb3 commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -1418,6 +1481,18 @@ func writeUint64(b *bytes.Buffer, v uint64) {
+// the browser (Teleport never receives this message, so won't be decoding it)
+const tdpMaxNotificationMessageLength = 10240
+
+// TODO: tdpMaxPathLength is currently just an arbitrary value for the maximum length of
```
Can we resolve this prior to merging to avoid creating more follow up work?

ibeckermayer replied:
This is an existing issue (https://github.com/gravitational/teleport/issues/14950 created July 27). May as well just figure it out here though, since otherwise it seems like one of those things that will just keep being de-prioritized.

ibeckermayer replied:
https://github.com/gravitational/teleport/issues/14950#issuecomment-1341632465

zmb3 commented on lib/srv/desktop/tdp/proto_test.go:
```diff
@@ -200,3 +188,165 @@ func TestMFA(t *testing.T) {
+			require.NoError(t, err)
+			_, err = Decode(bytes)
+			require.True(t, trace.IsLimitExceeded(err))
+			if !test.fatal {
+				require.True(t, IsNonFatalErr(err))
+				require.False(t, IsFatalErr(err))
+			} else {
+				require.False(t, IsNonFatalErr(err))
+				require.True(t, IsFatalErr(err))
+			}
```
```suggestion
			require.Equal(t, test.fatal, IsFatalErr(err))
			require.Equal(t, !test.fatal, IsNonFatalErr(err))
```

zmb3 commented on lib/web/desktop.go:
```diff
@@ -496,3 +507,18 @@ func (h *Handler) desktopAccessScriptInstallADCSHandle(w http.ResponseWriter, r
+
+// sendTDPNotification sends a tdp Notification over the supplied websocket with the
+// error message of err.
+func sendTDPNotification(ws *websocket.Conn, err error, fatal bool) error {
```
What if we just passed in a `Severity` instead of the `fatal` boolean? It would be less custom code here, and it would be obvious at the call site what the last argument is, instead of wondering what the bool is used for.

Commit:
CR

Commit:
Merge remote-tracking branch 'origin/master' into isaiah/fix-TEL-Q322-3

zmb3 commented on lib/srv/desktop/tdp/conn.go:
```diff
@@ -117,3 +118,27 @@ func (c *Conn) LocalAddr() net.Addr {
+		return false
+	}
+
+	return errors.Is(err, trace.LimitExceeded(clipDataMaxLenErrMsg)) ||
```
This is a pretty inefficient check, as you have to construct (up to) 4 errors for comparison, and each of the 4 `LimitExceeded` calls is going to compute the current stack trace.

WDYT about makign these sentinel errors at the package level?

```
var (
  mfaLimitExceededErr       = trace.LimitExceeded("asdfjkl")
  stringLenLimitExceededErr = trace.LimitExceeded("jkllp")
```

And then just checking `errors.Is` against those here.

ibeckermayer replied:
77cc8fc8dde8178d9706aaa1e23458a89f982b1d

zmb3 reviewed (commented):
One more suggestion on the error handling.

Commit:
fixes comment

LKozlowski commented on lib/srv/desktop/rdp/rdpclient/src/lib.rs:
```diff
@@ -1980,3 +1980,10 @@ pub(crate) type Messages = Vec<Message>;
+///
+/// Note that this is not an RDP defined value, but rather one we've chosen
+/// in order to harden system security.
+const MAX_ALLOWED_VCHAN_MSG_SIZE: usize = 2_usize.pow(21); // 2MB
```
nitpick: I know that it is the correct value, but at first for me it was a bit confusing at how it's the 2 to the power of 21 that makes 2MB.

Maybe it would be more readable if we separate the unit with the value, e.g:
```
const MAX_ALLOWED_VCHAN_MSG_SIZE: usize = 2 * 2_usize.pow(20);
```
or write it like:

```
const MAX_ALLOWED_VCHAN_MSG_SIZE: usize = 2 << 20;
```
or
```
const MAX_ALLOWED_VCHAN_MSG_SIZE: usize = 2 * 1024 * 1024;
```
It seems like we're using the last style in the codebase:
https://github.com/gravitational/teleport/blob/master/lib/events/sizelimit.go#L29

Of course, it's up to you! 😄

ibeckermayer replied:
77cc8fc8dde8178d9706aaa1e23458a89f982b1d

LKozlowski commented on lib/srv/desktop/rdp/rdpclient/src/vchan.rs:
```diff
@@ -27,24 +27,30 @@ use std::io::{Cursor, Read};
 /// See https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpbcgr/343e4888-4c48-4054-b0e3-4e0762d1993c
 /// for more information about chunks.
 pub struct Client {
+    // size_limit is the maximum size of an RDP
```
nitpick: can we clarify that's the size limit is in the bytes?

ibeckermayer replied:
77cc8fc8dde8178d9706aaa1e23458a89f982b1d

Commit:
CR

Commit:
Merge branch 'master' into isaiah/fix-TEL-Q322-3

Commit:
Merge branch 'master' into isaiah/fix-TEL-Q322-3

Commit:
Merge branch 'master' into isaiah/fix-TEL-Q322-3
